### PR TITLE
gRPC and CLI Command for Searching Requests from Database

### PR DESCRIPTION
**Pull request by ntchjb**
Fixed: [BAND-240](https://bandprotocol.atlassian.net/browse/BAND-240), [BAND-242](https://bandprotocol.atlassian.net/browse/BAND-242)

## Implementation details

Flag `--with-request-search` can be used to created a database that help collecting oracle requests from BandChain node. The syntax is as follow.

`--with-request-search=<sql-type>:<sql-data-source-name>`

Examples:

SQLite: `--with-request-search=sqlite3:/path/to/database-file.db`
PostgreSQL: `--with-request-search=postgres : host=localhost port=5432 user=postgres password=secret dbname=postgres sslmode=disable`
MySQL: `--with-request-search=mysql: user:password@tcp(localhost:5555)/dbname?tls=skip-verify&autocommit=true`

In order to define number of records for each particular search results, `--request-search-cache-size` flag with number of records specified can be used.

e.g. `--request-search-cache-size=5` means for each particular search query, there will be at most 5 latest oracle requests for that search query stored in database. So, if there are 5 possible search queries, there will be total of 25 records in database.

GORM is used to interact with database. the implementation now support MySQL, PostgreSQL, SQLite.

Once an oracle request has been resolved, all information of the request will be stored in database with indexing of following properties
- Oracle script ID
- Calldata
- Min Count
- Ask Count
- Resolve time (for ordering)

Then, request information in database can be queried by sending properties of the request.
- Oracle script ID
- Calldata
- Min Count
- Ask Count

## Please ensure the following requirements are met before submitting a pull request:

- [ ] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

**Review by Benzbeeb — APPROVED** [review]

File: proto/oracle/v1/query.proto [hunk 1]
```diff
@@ -68,10 +68,10 @@ service Query {
     option (google.api.http).get = "/oracle/v1/params";
   }
 
-  // RequestSearch queries the latest request that match search criteria.
-  rpc RequestSearch(QueryRequestSearchRequest)
-      returns (QueryRequestSearchResponse) {
-    option (google.api.http).get = "/oracle/v1/request_search";
+  // LatestRequest queries the latest request that match search criteria.
```
Comment (songwongtp) [inline proto/oracle/v1/query.proto:71]: I prefer "RequestSearch" more

File: x/oracle/client/cli/query.go [hunk 2]
```diff
@@ -187,21 +187,52 @@ func GetQueryCmdRequest() *cobra.Command {
...
+		Use:   "latest-request [oracle-script-id] [calldata-hex] [ask-count] [min-count]",
+		Short: "Query latest request based on given properties",
+		Args:  cobra.ExactArgs(4),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx, err := client.GetClientQueryContext(cmd)
+			if err != nil {
+				return err
+			}
+
+			oracleScriptID, err := strconv.ParseUint(args[0], 10, 64)
```
Comment (taobun) [inline x/oracle/client/cli/query.go:201]: strconv.ParseInt?

File: hooks/request/types.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+	// Reports
+	for _, report := range reports {
+		dbRequest.Reports = append(dbRequest.Reports, GenerateReportModel(result.RequestID, report))
+	}
+
+	return dbRequest
+}
+
+// GenerateRequestModel converts proto's Report to GORM's Report model
```
Comment (taobun) [inline hooks/request/types.go:231]: GenerateReportModel?

File: proto/oracle/v1/query.proto (hunk 1 above)
Comment (ntchjb) [reply proto/oracle/v1/query.proto]: This will be also fixed in #125 .

File: hooks/request/request.go [hunk 4]
```diff
@@ -83,62 +54,118 @@ func (h *Hook) AfterInitChain(ctx sdk.Context, req abci.RequestInitChain, res ab
...
+
+		switch event.Type {
+		case types.EventTypeReport:
+			reqID := types.RequestID(common.Atoi(evMap[types.EventTypeReport+"."+types.AttributeKeyID][0]))
+			validator := evMap[types.EventTypeReport+"."+types.AttributeKeyValidator][0]
+			iter := h.oracleKeeper.GetReportIterator(ctx, reqID)
+			defer iter.Close()
+
+			// Check all reports related to given request ID
+			for ; iter.Valid(); iter.Next() {
```
Comment (taobun) [inline hooks/request/request.go:76]: I think we can write `GetReport` function in keeper to find a report on request id and validator.

**Review by taobun — CHANGES_REQUESTED** [review]

File: hooks/request/types.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+	if len(r.IBCChannelID) > 0 {
+		ibcChannel = &types.IBCChannel{
+			PortId:    r.IBCPortID,
+			ChannelId: r.IBCChannelID,
+		}
+	}
+
+	// The whole response
+	request := types.QueryRequestResponse{
+		Request: &types.Request{
```
Comment (taobun) [inline hooks/request/types.go:145]: I think we using NewRequest to prevent a bug from missing some fields

File: hooks/request/db.go [hunk 6]
```diff
@@ -1,44 +1,123 @@
...
+
+func (h *Hook) insertReports(reportMap map[types.RequestID][]types.Report) {
+	var results []Report
+	for requestID, reports := range reportMap {
+		for _, report := range reports {
+			results = append(results, GenerateReportModel(requestID, report))
+		}
+	}
+
+	h.trans.Model(&Report{}).Create(&results)
```
Comment (taobun) [inline hooks/request/db.go:70]: Is there a case that we cannot insert new report in DB? (request of this report has been deleted from DB)

**Review by taobun — CHANGES_REQUESTED** [review]

File: hooks/request/types.go (hunk 5 above)
Comment (ntchjb) [reply hooks/request/types.go]: I have added a condition to check whether request exists or not.

File: hooks/request/db.go (hunk 6 above)
Comment (ntchjb) [reply hooks/request/db.go:72]: I have added a condition to check whether request exists or not.

File: hooks/request/types.go (hunk 5 above)
Comment (ntchjb) [reply hooks/request/types.go]: Fixed. It's now using function to create new instances.

File: hooks/request/request.go [hunk 7]
```diff
@@ -83,62 +54,124 @@ func (h *Hook) AfterInitChain(ctx sdk.Context, req abci.RequestInitChain, res ab
...
+				ctx.Logger().Error("Unable to parse validator address got from EventTypeReport for request search", "error", err)
+				continue
+			}
+			report, err := h.oracleKeeper.GetReport(ctx, reqID, valAddr)
+			if err != nil {
+				ctx.Logger().Error("Unable to get report for request search", "error", err)
+				continue
+			}
+			// Collect reports which are submitted AFTER the request successfully resolved
+			if !report.InBeforeResolve && report.Validator == validator {
```
Comment (taobun) [inline hooks/request/request.go:83]: Remove `report.Validator == validator`

**Review by taobun — CHANGES_REQUESTED** [review]

File: hooks/request/request.go [hunk 8]
```diff
@@ -83,62 +54,124 @@ func (h *Hook) AfterInitChain(ctx sdk.Context, req abci.RequestInitChain, res ab
...
+			queryReq.Calldata,
+			queryReq.AskCount,
+			queryReq.MinCount,
+		)
+
+		// check query results
+		var qReqRes *types.QueryRequestResponse
+		if err != nil {
+			if errors.Is(err, gorm.ErrRecordNotFound) {
+				qReqRes = nil
```
Comment (taobun) [inline hooks/request/request.go:156]: Should we return error if not found?

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]
LGTM

**Review by Benzbeeb — APPROVED** [review]

**Pull request merged**